ДЗ 5

valenok95 opened the pull request.

Commit:
ДЗ 5

vlad-iv commented on homework05/src/main/java/com/example/homework05/controller/ShellController.java:
```diff
@@ -0,0 +1,66 @@
+            @ShellOption(defaultValue = "Some book") String name,
+            @ShellOption(defaultValue = "1") long authorId,
+            @ShellOption(defaultValue = "1") long genreId) {
+        Author author = authorService.getAuthorById(authorId);
```
Предлагаю логику перенести в сервис
т.е. bookService.saveBook(BookDto newBook)
в newBook находятся все данные введенные  с Shell
Так же и остальные методы шела

vlad-iv commented on homework05/src/main/java/com/example/homework05/controller/ShellController.java:
```diff
@@ -0,0 +1,66 @@
+@RequiredArgsConstructor
+public class ShellController {
+    private final BookService bookService;
+    private final AuthorService authorService;
```
можно удалить

vlad-iv commented on homework05/src/main/java/com/example/homework05/controller/ShellController.java:
```diff
@@ -0,0 +1,66 @@
+public class ShellController {
+    private final BookService bookService;
+    private final AuthorService authorService;
+    private final GenreService genreService;
```
можно удалить

vlad-iv commented on homework05/src/main/java/com/example/homework05/dao/AuthorDao.java:
```diff
@@ -0,0 +1,16 @@
+public interface AuthorDao {
+    int count();
+
+    void insert(Author author);
```
Предлагаю реализовать сигнатуру 
Author insert(Author author)

vlad-iv commented on homework05/src/main/java/com/example/homework05/dao/AuthorDao.java:
```diff
@@ -0,0 +1,16 @@
+
+    void insert(Author author);
+
+    Author getById(long id);
```
Желательно реализовать сигнатуру
Optional< Author  > getById(long id);

vlad-iv commented on homework05/src/main/java/com/example/homework05/dao/BookDao.java:
```diff
@@ -0,0 +1,16 @@
+import com.example.homework05.domain.Book;
+import java.util.List;
+
+public interface BookDao {
```
Предлагаю применить те же рекомендации что для AuthorDao

vlad-iv commented on homework05/src/main/java/com/example/homework05/dao/GenreDao.java:
```diff
@@ -0,0 +1,16 @@
+import com.example.homework05.domain.Genre;
+import java.util.List;
+
+public interface GenreDao {
```
Предлагаю применить те же рекомендации что для AuthorDao

vlad-iv commented on homework05/src/main/java/com/example/homework05/domain/Author.java:
```diff
@@ -0,0 +1,11 @@
+@RequiredArgsConstructor
+@Data
+public class Author {
+    private final long id;
```
Long id
так как при добавлении, у этого поля не будет значения, мы его зададим

vlad-iv commented on homework05/src/main/java/com/example/homework05/domain/Book.java:
```diff
@@ -0,0 +1,13 @@
+@RequiredArgsConstructor
+@Data
+public class Book {
+    private final long id;
```
Long id

vlad-iv commented on homework05/src/main/java/com/example/homework05/domain/Genre.java:
```diff
@@ -0,0 +1,11 @@
+@RequiredArgsConstructor
+@Data
+public class Genre {
+    private final long id;
```
Long id

vlad-iv commented on homework05/src/main/java/com/example/homework05/service/AuthorService.java:
```diff
@@ -0,0 +1,25 @@
+public class AuthorService {
+    private final AuthorDao authorDao;
+
+    public void saveAuthor(Author author) {
```
Предлагаю реализовать сигнатуру 
public Author saveAuthor(Author author)

vlad-iv commented on homework05/src/main/java/com/example/homework05/service/BookService.java:
```diff
@@ -0,0 +1,24 @@
+    public void saveBook(Book book) {
+        bookDao.insert(book);
+    }
+
```
Предлагаю реализовать обновление книги

vlad-iv commented on homework05/src/main/java/com/example/homework05/service/BookService.java:
```diff
@@ -0,0 +1,24 @@
+public class BookService {
+    private final BookDao bookDao;
+
+    public void saveBook(Book book) {
```
BookDto  saveBook(BookDto book)

vlad-iv commented on homework05/src/main/java/com/example/homework05/service/BookService.java:
```diff
@@ -0,0 +1,24 @@
+    }
+
+    public Book getBookById(long id) {
+        return bookDao.getById(id);
```
если книга не найдена, ожидается, что будет исключение выброшено в сервисе
(сейчас в Dao)

vlad-iv commented on homework05/src/main/resources/schema.sql:
```diff
@@ -0,0 +1,9 @@
+DROP TABLE IF EXISTS BOOKS;
```
предлагаю удалить DROP TABLE

vlad-iv commented on homework05/src/main/resources/schema.sql:
```diff
@@ -0,0 +1,9 @@
+DROP TABLE IF EXISTS BOOKS;
+DROP TABLE IF EXISTS AUTHORS;
+DROP TABLE IF EXISTS GENRES;
+CREATE TABLE AUTHORS (ID BIGINT PRIMARY KEY, name VARCHAR(255));
```
предлагаю добавить if not exit

vlad-iv commented on homework05/src/main/resources/schema.sql:
```diff
@@ -0,0 +1,9 @@
+DROP TABLE IF EXISTS GENRES;
+CREATE TABLE AUTHORS (ID BIGINT PRIMARY KEY, name VARCHAR(255));
+CREATE TABLE GENRES(ID BIGINT PRIMARY KEY, name VARCHAR(255));
+CREATE TABLE BOOKS(ID BIGINT PRIMARY KEY, name VARCHAR(255),
```
Предлагаю придерживаться нейминга полей для БД
AUTHOR_ID 
через нижнее подчеркиваение

vlad-iv commented on homework05/src/main/resources/schema.sql:
```diff
@@ -0,0 +1,9 @@
+DROP TABLE IF EXISTS AUTHORS;
+DROP TABLE IF EXISTS GENRES;
+CREATE TABLE AUTHORS (ID BIGINT PRIMARY KEY, name VARCHAR(255));
+CREATE TABLE GENRES(ID BIGINT PRIMARY KEY, name VARCHAR(255));
```
рекомендуется вместо ID называть колонку идентификатора имя сущности + ид
GENRE_ID

Commit:
Исправлены замечания:
1. сигнатуры DAO getById(long id) возвращают Optional
2. сигнатуры DAO insert() возвращают возвращают сохраненный объект
3. переименованы поля в DB : СУЩНОСТЬ_ID
4. На уровне контроллера в сервис передаётся только Dto

vlad-iv commented on homework05/build.gradle:
```diff
@@ -0,0 +1,32 @@
+    implementation 'org.springframework.boot:spring-boot-starter-jdbc'
+    implementation group: 'org.springframework.shell', name:'spring-shell-starter',  version: '3.1.3'
+    implementation 'com.h2database:h2'
+    compileOnly group: 'org.projectlombok', name: 'lombok', version: '1.18.26'
```
можно не указывать версию для lombok

vlad-iv commented on homework05/src/main/java/com/example/homework05/service/BookService.java:
```diff
@@ -0,0 +1,37 @@
+@RequiredArgsConstructor
+public class BookService {
+    private final BookDao bookDao;
+    private final AuthorService authorService;
```
Предлагаю заменить на dao
ожидается, что сервисы не зависят от других сервисов сущностей

vlad-iv commented on homework05/src/main/java/com/example/homework05/service/BookService.java:
```diff
@@ -0,0 +1,37 @@
+public class BookService {
+    private final BookDao bookDao;
+    private final AuthorService authorService;
+    private final GenreService genreService;
```
Предлагаю заменить на dao

Commit:
1. Добавлен метод update для книг в DAO и Service
2. Добавлены тесты

Commit:
1. Добавлен метод update для книг в DAO и Service
2. Добавлены тесты
3. в BookService только DAO без зависимости от других сервисов

vlad-iv commented on homework05/src/main/java/com/example/homework05/service/BookService.java:
```diff
@@ -0,0 +1,47 @@
+    public void saveBook(BookDto newBook) {
+        Author author = authorDao.getById(newBook.getAuthorId()).orElseThrow();
+        Genre genre = genreDao.getById(newBook.getGenreId()).orElseThrow();
+        Book book = new Book(newBook.getId(), newBook.getName(),
```
У новой книги нет id
предлагаю указать null

vlad-iv commented on homework05/src/main/java/com/example/homework05/service/BookService.java:
```diff
@@ -0,0 +1,47 @@
+    }
+
+    public void updateBook(BookDto updateBook) {
+        Author author = authorDao.getById(updateBook.getAuthorId()).orElseThrow();
```
Предлагаю выбросить свое исключение NotFoundException

vlad-iv commented on homework05/src/main/java/com/example/homework05/service/BookService.java:
```diff
@@ -0,0 +1,47 @@
+    public void updateBook(BookDto updateBook) {
+        Author author = authorDao.getById(updateBook.getAuthorId()).orElseThrow();
+        Genre genre = genreDao.getById(updateBook.getGenreId()).orElseThrow();
+        Book book = new Book(updateBook.getId(), updateBook.getName(),
```
Осталось проверить что с таким ид есть есть книга, и если нет то выбросить NotFoundException

Commit:
1. автогенератор id
2. своё исключение на ошибку получения книги
